OrbitCore/OrbitModule.cpp

```diff
@@ -108,45 +107,62 @@ Function* Pdb::FunctionFromName(const std::string& a_Name) {
 }
 
 //-----------------------------------------------------------------------------
-std::string FindSymbols(const std::string& module_path) {
+std::unique_ptr<ElfFile> FindSymbols(std::string& module_path) {
+  auto module_elf_file = ElfFile::Create(module_path);
+  if (module_elf_file == nullptr) {
+    PRINT(absl::StrFormat("Unable to load module (object file) from \"%s\"",
+                          module_path));
+    return nullptr;
+  }
+
+  if (module_elf_file->ContainsSymbols()) {
+    return module_elf_file;
+  }
+
   // Look for .debug files associated with passed in module name
-  // TODO: .debug file might not have same name as module, we should check
-  //       for unique identifier in the symbols file...
+  // TODO: .debug file might not have same name as module, maybe there is a good
+  //       way of searching files based on buildId
 
   std::string dir = Path::GetDirectory(module_path);
-  std::vector<std::string> symbolDirectories = {"~/", Path::GetHome(), dir,
-                                                dir + "debug_symbols/", "/home/cloudcast/"};
-
   std::string file = Path::StripExtension(Path::GetFileName(module_path));
 
-  for (auto& symbolDirectory : symbolDirectories) {
-    std::string debugFile = symbolDirectory + file + ".debug";
-    if (Path::FileExists(debugFile)) {
-      PRINT_VAR(debugFile);
-      return debugFile;
+  std::vector<std::string> searchDirectories = {
+      dir, dir + "debug_symbols/", Path::GetHome(), "/home/cloudcast/"};
+  std::vector<std::string> symbolsFileEndings = {".debug", ".elf.debug"};
+
+  std::vector<std::string> searchFilePaths;
+  for (const auto& directory : searchDirectories) {
+    for (const auto& fileEnding : symbolsFileEndings) {
+      searchFilePaths.emplace_back(directory + file + fileEnding);
     }
+  }
+
+  for (const auto& symbolsFilePath : searchFilePaths) {
+    if (!Path::FileExists(symbolsFilePath)) continue;
 
-    debugFile = symbolDirectory + file + ".elf.debug";
-    if (Path::FileExists(debugFile)) {
-      PRINT_VAR(debugFile);
-      return debugFile;
+    auto symbolsFile = ElfFile::Create(symbolsFilePath);
+    if (symbolsFile == nullptr) continue;
+
+    if (!symbolsFile->ContainsSymbols()) continue;
+
+    const std::string& moduleBuildId(module_elf_file->GetBuildId());
+    if (moduleBuildId != "" && moduleBuildId == symbolsFile->GetBuildId()) {
+      module_path = symbolsFilePath;
+      return symbolsFile;
     }
   }
 
-  return module_path;
+  return nullptr;
 }
 
-Pdb::~Pdb()
-{
-  PRINT_FUNC;
-}
+Pdb::~Pdb() { PRINT_FUNC; }
 
 bool Pdb::LoadFunctions(const char* file_name) {
-  m_FileName = FindSymbols(file_name);
-  m_Name = Path::GetFileName(m_FileName);
   m_LoadedModuleName = file_name;
+  m_FileName = file_name;
+  auto elf_file = FindSymbols(m_FileName);
```

Comment:
Once again, let's avoid auto except for things like for loops (improves readability), assignments from explicit casts, make_shared/make_unique (prevents specifying the type twice). Same for the next commit.